### ISSUE TITLE
glibc-2.27 memfd_create not being detected correctly with kernels 3.10.y/3.14.y

### ISSUE THREAD

**Issue by MilhouseVH** [systemd/systemd#8099]
### Submission type

- Bug report

### systemd version the issue has been seen with

> 237

### Used distribution

>LibreELEC 9.0

### In case of bug report: Expected behaviour you didn't see

>Successful systemd build with glibc-2.27 and kernels that do not implement `memfd_create`.

### In case of bug report: Unexpected behaviour you saw

>systemd build failure with glibc-2.27 and kernels that do not implement `memfd_create`.

> systemd builds successfully with glibc-2.27 and kernel 4.14.16.

### In case of bug report: Steps to reproduce the problem

> Build systemd-237 with glibc-2.27 and kernels 3.10.108 or 3.14.29 (most likely any other kernel prior to 3.17.0 when the `memfd_create` syscall is introduced to the kernel).

I've encountered a problem building systemd-237 and glibc-2.27 with kernels that do not implement the `memfd_create` syscall (ie. any kernel before 3.17.0).

For example, when building 3.10.108, glibc-2.27 and systemd-237:
```
[66/985] Compiling C object 'src/basic/basic@sta/barrier.c.o'.
[67/985] Compiling C object 'src/basic/basic@sta/calendarspec.c.o'.
FAILED: src/basic/basic@sta/calendarspec.c.o
/home/neil/projects/scratch/alternates/LibreELEC.tv/build.LibreELEC-WeTek_Hub.arm-9.0-devel-next/toolchain/bin/armv8a-libreelec-linux-gnueabi-gcc -march=armv8-a+crc -mabi=aapcs-linux -Wno-psabi -Wa,-mno-warn-deprecated -mcpu=cortex-a53 -mfloat-abi=hard -mfpu=neon-fp-armv8 -fomit-frame-pointer -Wall -pipe -Os -flto -ffat-lto-objects -Isrc/basic/basic@sta -Isrc/basic -I../src/basic -I. -I../ -Isrc/libsystemd-network -I../src/libsystemd-network -I../src/libsystemd/sd-network -I../src/libsystemd/sd-netlink -I../src/libsystemd/sd-id128 -I../src/libsystemd/sd-hwdb -I../src/libsystemd/sd-device -I../src/libsystemd/sd-bus -Isrc/core -I../src/core -Isrc/libudev -I../src/libudev -Isrc/udev -I../src/udev -Isrc/login -I../src/login -Isrc/timesync -I../src/timesync -Isrc/resolve -I../src/resolve -Isrc/journal -I../src/journal -Isrc/systemd -I../src/systemd -Isrc/shared -I../src/shared -I/home/neil/projects/scratch/alternates/LibreELEC.tv/build.LibreELEC-WeTek_Hub.arm-9.0-devel-next/toolchain/armv8a-libreelec-linux-gnueabi/sysroot/usr/include -I/home/neil/projects/scratch/alternates/LibreELEC.tv/build.LibreELEC-WeTek_Hub.arm-9.0-devel-next/toolchain/armv8a-libreelec-linux-gnueabi/sysroot/usr/include/blkid -I/home/neil/projects/scratch/alternates/LibreELEC.tv/build.LibreELEC-WeTek_Hub.arm-9.0-devel-next/toolchain/armv8a-libreelec-linux-gnueabi/sysroot/usr/include/uuid -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -std=gnu99 -Wextra -Werror=undef -Wlogical-op -Wmissing-include-dirs -Wold-style-definition -Wpointer-arith -Winit-self -Wdeclaration-after-statement -Wfloat-equal -Wsuggest-attribute=noreturn -Werror=missing-prototypes -Werror=implicit-function-declaration -Werror=missing-declarations -Werror=return-type -Werror=incompatible-pointer-types -Werror=format=2 -Wstrict-prototypes -Wredundant-decls -Wmissing-noreturn -Wimplicit-fallthrough=5 -Wshadow -Wendif-labels -Wstrict-aliasing=2 -Wwrite-strings -Werror=overflow -Wdate-time -Wnested-externs -ffast-math -fno-common -fdiagnostics-show-option -fno-strict-aliasing -fvisibility=hidden -fstack-protector -fstack-protector-strong --param=ssp-buffer-size=4 -fPIE -Wno-unused-parameter -Wno-missing-field-initializers -Wno-unused-result -Wno-format-signedness -Wno-error=nonnull -Werror=shadow -ffunction-sections -fdata-sections -include config.h -fPIC -pthread -fvisibility=default -MMD -MQ 'src/basic/basic@sta/calendarspec.c.o' -MF 'src/basic/basic@sta/calendarspec.c.o.d' -o 'src/basic/basic@sta/calendarspec.c.o' -c ../src/basic/calendarspec.c
In file included from ../src/basic/missing.h:1358:0,
                 from ../src/basic/util.h:47,
                 from ../src/basic/calendarspec.h:29,
                 from ../src/basic/calendarspec.c:34:
../src/basic/missing_syscall.h:65:19: error: static declaration of 'memfd_create' follows non-static declaration
 static inline int memfd_create(const char *name, unsigned int flags) {
                   ^~~~~~~~~~~~
In file included from /home/neil/projects/scratch/alternates/LibreELEC.tv/build.LibreELEC-WeTek_Hub.arm-9.0-devel-next/toolchain/armv8a-libreelec-linux-gnueabi/sysroot/usr/include/bits/mman-linux.h:115:0,
                 from /home/neil/projects/scratch/alternates/LibreELEC.tv/build.LibreELEC-WeTek_Hub.arm-9.0-devel-next/toolchain/armv8a-libreelec-linux-gnueabi/sysroot/usr/include/bits/mman.h:40,
                 from /home/neil/projects/scratch/alternates/LibreELEC.tv/build.LibreELEC-WeTek_Hub.arm-9.0-devel-next/toolchain/armv8a-libreelec-linux-gnueabi/sysroot/usr/include/sys/mman.h:41,
                 from ../src/basic/calendarspec.c:30:
/home/neil/projects/scratch/alternates/LibreELEC.tv/build.LibreELEC-WeTek_Hub.arm-9.0-devel-next/toolchain/armv8a-libreelec-linux-gnueabi/sysroot/usr/include/bits/mman-shared.h:46:5: note: previous declaration of 'memfd_create' was here
 int memfd_create (const char *__name, unsigned int __flags) __THROW;
     ^~~~~~~~~~~~
[68/985] Compiling C object 'src/basic/basic@sta/cap-list.c.o'.
[69/985] Compiling C object 'src/basic/basic@sta/capability-util.c.o'.
[70/985] Compiling C object 'src/basic/basic@sta/clock-util.c.o'.
[71/985] Compiling C object 'src/basic/basic@sta/exit-status.c.o'.
[72/985] Compiling C object 'src/basic/basic@sta/ether-addr-util.c.o'.
[73/985] Compiling C object 'src/basic/basic@sta/copy.c.o'.
[74/985] Compiling C object 'src/basic/basic@sta/btrfs-util.c.o'.
[75/985] Compiling C object 'src/basic/basic@sta/extract-word.c.o'.
[76/985] Compiling C object 'src/basic/basic@sta/cgroup-util.c.o'.
../src/basic/cgroup-util.c: In function 'cg_pid_get_path':
../src/basic/cgroup-util.c:1066:48: warning: 'controller_str' may be used uninitialized in this function [-Wmaybe-uninitialized]
                                 if (k == cs && memcmp(word, controller_str, cs) == 0) {
                                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ninja: build stopped: subcommand failed.
```

The `HAVE_MEMFD_CREATE` detection in `meson.build` is failing to determine that `memfd_create` is now implemented by glibc-2.27, however it appears to fail only when the kernel does not also implement this syscall. If the kernel does implement `memfd_create`, ie. 4.14.16, then the meson detection works as expected and the build succeeds.

system-237 sets the following `#define`s when building with glibc-2.27 and the stated kernels:

kernel 3.10.108:
```
config.h:#define HAVE_LINUX_MEMFD_H 0
config.h:#define HAVE_MEMFD_CREATE 0
```

kernel 3.14.29:
```
config.h:#define HAVE_LINUX_MEMFD_H 0
config.h:#define HAVE_MEMFD_CREATE 0
```

kernel 4.14.16:
```
config.h:#define HAVE_LINUX_MEMFD_H 1
config.h:#define HAVE_MEMFD_CREATE 1
```

Consequently systemd-237 fails to build with glibc-2.27 and kernels 3.10.108 and 3.14.29 (see build error, above), but builds successfully with 4.14.16.

Since LibreELEC will only be building with `glibc-2.27` (which implements `memfd_create`) as a workaround I'm using [this patch](https://github.com/MilhouseVH/LibreELEC.tv/blob/4579822797b1575fd064d87bcf33131511f9e471/packages/sysutils/systemd/patches/systemd-0100-fix-glibc-2.27-build.patch) to force enable `HAVE_MEMFD_CREATE` in `meson.build` and I'm able to  successfully build systemd-237 with all kernels.

The patch confirms that systemd-237 is building with the glibc-2.27 implementation of `memfd_create` because if we use the patch and build systemd-237 with glibc-2.26 (and not glibc-2.27) then systemd will fail to build as `memfd_create` is no longer declared. For example:
```
FAILED: src/basic/basic@sta/fileio.c.o
/home/neil/projects/scratch/alternates/LibreELEC.tv/build.LibreELEC-WeTek_Hub.arm-9.0-devel-next/toolchain/bin/armv8a-libreelec-linux-gnueabi-gcc -march=armv8-a+crc -mabi=aapcs-linux -Wno-psabi -Wa,-mno-warn-deprecated -mcpu=cortex-a53 -mfloat-abi=hard -mfpu=neon-fp-armv8 -fomit-frame-pointer -Wall -pipe -Os -flto -ffat-lto-objects -Isrc/basic/basic@sta -Isrc/basic -I../src/basic -I. -I../ -Isrc/libsystemd-network -I../src/libsystemd-network -I../src/libsystemd/sd-network -I../src/libsystemd/sd-netlink -I../src/libsystemd/sd-id128 -I../src/libsystemd/sd-hwdb -I../src/libsystemd/sd-device -I../src/libsystemd/sd-bus -Isrc/core -I../src/core -Isrc/libudev -I../src/libudev -Isrc/udev -I../src/udev -Isrc/login -I../src/login -Isrc/timesync -I../src/timesync -Isrc/resolve -I../src/resolve -Isrc/journal -I../src/journal -Isrc/systemd -I../src/systemd -Isrc/shared -I../src/shared -I/home/neil/projects/scratch/alternates/LibreELEC.tv/build.LibreELEC-WeTek_Hub.arm-9.0-devel-next/toolchain/armv8a-libreelec-linux-gnueabi/sysroot/usr/include -I/home/neil/projects/scratch/alternates/LibreELEC.tv/build.LibreELEC-WeTek_Hub.arm-9.0-devel-next/toolchain/armv8a-libreelec-linux-gnueabi/sysroot/usr/include/blkid -I/home/neil/projects/scratch/alternates/LibreELEC.tv/build.LibreELEC-WeTek_Hub.arm-9.0-devel-next/toolchain/armv8a-libreelec-linux-gnueabi/sysroot/usr/include/uuid -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -std=gnu99 -Wextra -Werror=undef -Wlogical-op -Wmissing-include-dirs -Wold-style-definition -Wpointer-arith -Winit-self -Wdeclaration-after-statement -Wfloat-equal -Wsuggest-attribute=noreturn -Werror=missing-prototypes -Werror=implicit-function-declaration -Werror=missing-declarations -Werror=return-type -Werror=incompatible-pointer-types -Werror=format=2 -Wstrict-prototypes -Wredundant-decls -Wmissing-noreturn -Wimplicit-fallthrough=5 -Wshadow -Wendif-labels -Wstrict-aliasing=2 -Wwrite-strings -Werror=overflow -Wdate-time -Wnested-externs -ffast-math -fno-common -fdiagnostics-show-option -fno-strict-aliasing -fvisibility=hidden -fstack-protector -fstack-protector-strong --param=ssp-buffer-size=4 -fPIE -Wno-unused-parameter -Wno-missing-field-initializers -Wno-unused-result -Wno-format-signedness -Wno-error=nonnull -Werror=shadow -ffunction-sections -fdata-sections -include config.h -fPIC -pthread -fvisibility=default -MMD -MQ 'src/basic/basic@sta/fileio.c.o' -MF 'src/basic/basic@sta/fileio.c.o.d' -o 'src/basic/basic@sta/fileio.c.o' -c ../src/basic/fileio.c
../src/basic/fileio.c: In function 'open_serialization_fd':
../src/basic/fileio.c:1460:14: error: implicit declaration of function 'memfd_create'; did you mean 'timer_create'? [-Werror=implicit-function-declaration]
         fd = memfd_create(ident, MFD_CLOEXEC);
              ^~~~~~~~~~~~
              timer_create
../src/basic/fileio.c:1460:14: warning: nested extern declaration of 'memfd_create' [-Wnested-externs]
cc1: some warnings being treated as errors
```

However this patch, while it works, is not ideal as we'd rather not patch against upstream. And it's just a hack.

Does anyone have any idea why `HAVE_MEMFD_CREATE` is not being determined correctly by `meson.build` when the kernel does not implement `memfd_create` but `glibc` does?

**Comment by yuwata** [comment]
If I understand correctly, `memfd_create()` is added to glibc quite recently
https://sourceware.org/git/?p=glibc.git;a=commit;h=59d2cbb1fe4b8601d5cbd359c3806973eab6c62d
So, for such system which does not have recent glibc, `HAVE_MEMFD_CREATE` is set to false.

Note that, in the case of Fedora, v27 (the most recent release) does not have `memfd_create()` in glibc header, and the current development (rawhide) has it.

**Comment by MilhouseVH** [comment]
Yes, `memfd_create` is added in glibc-2.27, which is the version LibreELEC 9.0 is building with, yet systemd-237 is failing to correctly determine that glibc-2.27 is now providing `memfd_create` when building with these older 3.x kernels (that do not implement `memfd_create`). Consequently the systemd build fails when systemd redeclares `memfd_create` which is being provided by the glibc-2.27 headers (the same glibc headers that meson.build failed to find during configure).

I believe https://github.com/systemd/systemd/commit/85db59b794d1ca7f16ea816c916bb4472958cc1b should have addressed this issue and should now be configuring `#define HAVE_MEMFD_CREATE 1` regardless of kernel version when `memfd_create` is provided by glibc-2.27, but for some reason it isn't working reliably and the test fails with these older 3.x kernels, meaning `HAVE_MEMFD_CREATE` remains disabled when it should be enabled.

In short, it would appear that `HAVE_MEMFD_CREATE` is not working with glibc-2.27 depending on the kernel (although this is just a guess, based on the pattern of failures I've seen so far with these 3 kernels).

**Comment by keszybz** [comment]
We don't have glibc-2.27 in Fedora. The latest is glibc-devel-2.26.9000-51.fc28.x86_64. I expect that it's essentially identical to the released version though.

I tried to replicate the build failure with glibc-2.26.9000, and kernels going back to kernel-headers-4.9.4-201.fc25.x86_64, and it builds fine. But I'm on amd64, arm might be different.

Please attach `build/meson-logs/meson-log.txt` from one of the affected systems.

**Comment by keszybz** [comment]
Builds fine on Fedora arm too: kernel-headers-4.14.13-300.fc27.armv7hl and glibc-devel-2.26.9000-51.fc28.armv7hl.

**Comment by MilhouseVH** [comment]
There is no problem with glibc-2.26 - I can build systemd-237 with glibc-2.26 and (for example) kernel 3.10.108 without any issue, as glibc-2.26 does not implement `memfd_create`.

However, when building with glibc-2.27 and systemd-237 and 3.10.108 (or 3.14.29), systemd will fail to build as systemd fails to determine the presence of `memfd_create` which is now being provided by glibc-2.27.

This only seems to be an issue with kernels that DON'T implement `memfd_create` (so anything prior to 3.17.0) - there is no problem with 4.14.16, but there is a problem with 3.10.108 and 3.14.29.

meson-log.txt (systemd-237 + glibc-2.27 + kernel 3.14.29): http://ix.io/FhH (which fails)
meson-log.txt (systemd-237 + glibc-2.27 + kernel 4.14.16): http://ix.io/FhJ (successful)

With 3.14.29, the first test fails as `__stub_memfd_create` is defined in `include/gnu/stubs-hard.h`, which is correct as this kernel doesn't implement the syscall.

The next test also fails (but shouldn't!) because `memfd_create` is not defined, however there is a `memfd_create` function declaration in the glibc-2.27 header - it's in [include/bits/mman-shared.h](http://ix.io/FhN).

Is the `! defined(memfd_create)` test correct? I'm not a C guru so I'm not sure if `defined` is a valid test for a function declaration. But this second test should be successful as glibc-2.27 is now providing `memfd_create`.

**Comment by keszybz** [comment]
Oh, I think it comes down to the meson version. With meson-0.44.0:
```
Code:
 #include <sys/mman.h>
#include <limits.h>

        #if defined __stub_memfd_create || defined __stub___memfd_create
        fail fail fail this function is not going to work
        #endif
        
int main() {
            void *a = (void*) &memfd_create;
            long b = (long) a;
            return (int) b;
        }
Compiler stdout:
 
Compiler stderr:
 
Checking for function "memfd_create": YES
```
Essentially, the test is broken in meson-0.42.

**Comment by MilhouseVH** [comment]
Ah OK thanks, I'll try upgrading to meson-0.44.0 and will report back once I've tested.

**Comment by MilhouseVH** [comment]
Unfortunately I'm not seeing any difference with meson-0.44.0: http://ix.io/Fik

Note that it's not the first test that is the problem, it's the second:
```
Running compile:
Working directory:  /tmp/tmp63a184_e
Command line:  /home/neil/projects/scratch/alternates/LibreELEC.tv/build.LibreELEC-WeTek_Hub.arm-9.0-devel-next/toolchain/bin/armv8a-libreelec-linux-gnueabi-gcc /tmp/tmp63a184_e/testfile.c -march=armv8-a+crc -mabi=aapcs-linux -Wno-psabi -Wa,-mno-warn-deprecated -mcpu=cortex-a53 -mfloat-abi=hard -mfpu=neon-fp-armv8 -fomit-frame-pointer -Wall -pipe -Os -flto -ffat-lto-objects -march=armv8-a+crc -mtune=cortex-a53 -Wl,--as-needed -fuse-ld=gold -fuse-linker-plugin -flto -O0 -std=gnu99 -D_GNU_SOURCE -D_FILE_OFFSET_BITS=64 -o /tmp/tmp63a184_e/output.exe 

Code:
 #include <sys/mman.h>
        int main() {
        #ifdef __has_builtin
            #if !__has_builtin(__builtin_memfd_create)
                #error "__builtin_memfd_create not found"
            #endif
        #elif ! defined(memfd_create)
            /* Check for __builtin_memfd_create only if no includes were added to the
             * prefix above, which means no definition of memfd_create can be found.
             * We would always check for this, but we get false positives on
             * MSYS2 if we do. Their toolchain is broken, but we can at least
             * give them a workaround. */
            #if 0
                __builtin_memfd_create;
            #else
                #error "No definition for __builtin_memfd_create found in the prefix"
            #endif
        #endif
        }
Compiler stdout:
 
Compiler stderr:
 /tmp/tmp63a184_e/testfile.c: In function 'main':
/tmp/tmp63a184_e/testfile.c:16:18: error: #error "No definition for __builtin_memfd_create found in the prefix"
                 #error "No definition for __builtin_memfd_create found in the prefix"
                  ^~~~~

Checking for function "memfd_create": NO
```
The failure of this second test leaves `HAVE_MEMFD_CREATE` set to 0, which results in systemd trying to redeclare the `memfd_create` syscall that is now implemented by glibc-2.27.

**Comment by keszybz** [comment]
Oh, OK. So the test fails because `__stub_memfd_create` or `__stub___memfd_create` is defined. This defeats the test which checks for function `memfd_create`. Then the second test checks for `__builtin_memfd_create`, which doesn't exist. So the real problem is the first fails, not the second.

But on Fedora, even with kernel-headers-3.16.3-302.fc21.x86_64, `memfd_create` is detected properly with new glibc. So it seems something is different in your glibc headers...

**Comment by MilhouseVH** [comment]
By the way I'm building with gcc-7.3.0. I don't know what else to look for, but I have a workaround for now. If not reproducible, then no big deal.

**Comment by keszybz** [comment]
It's probably reproducible, but requires the right combination of kernel/glibc...

**Comment by keszybz** [comment]
I now tested with glibc-devel-2.27-1.fc28.x86_64, and it works.

**Comment by MilhouseVH** [comment]
Did I mention I'm cross-compiling? :)

**Comment by keszybz** [comment]
To recap: the build is failing because either `__stub_memfd_create` or `__stub___memfd_create` is defined, so meson assumes that `stub_memfd_create` is not declared, but it actually is declared, so the build fails, because the system definition conflicts with our definition.

The real issue is that there's confusion in what `meson.get_compiler().has_function()` means:
- meson effectively checks if the function is **defined** and **usable**
- we need a check if the function is **defined**
So either our logic or meson needs to change.

**Comment by sonvirgo** [comment]
I am on systemd 238 and have the same issue while making pulseaudio-raop2.
Work around please.
Kali 2017.3

**Comment by floppym** [comment]
> I am on systemd 238 and have the same issue while making pulseaudio-raop2.

That's a problem you should report to the developers of "pulseaudio-raop2". It has nothing to do wit hsystemd.

**Comment by sonvirgo** [comment]
@floppym  Really? Becasue it makes fine with systemd 234, Ubuntu Zesty, so I suspect some thing wrong in systemd238?

**Comment by floppym** [comment]
Ubuntu Zesty probably has an older version of glibc.

**Comment by sonvirgo** [comment]
@floppym  I think it glibc 2.24-9ubuntu2.2

**Comment by floppym** [comment]
glibc-2.24 is obviously older than glibc-2.27, which is the version which triggered the bug reported in this systemd bug report.

Again, your pulseaudio-raop2 build failure has nothing to do with systemd, despite what your Google search may be returning.

**Comment by sonvirgo** [comment]
Ok Thanks!
By the way, as I am new to linux do you have any help to work around like patch or upgrade/downgrade glibc?

**Comment by floppym** [comment]
It will probably work if you remove the function definition from https://github.com/hfujita/pulseaudio-raop2/blob/hf/raop2-v7-next/src/pulsecore/memfd-wrappers.h#L36.

**Comment by sonvirgo** [comment]
@floppym Thank you! Although it is late, I will try your suggestion, tomorrow morning after the system finish upgrading  with glibc 2.7.2
Regards!

**Comment by sonvirgo** [comment]
Your suggestion work, however it not makes it due to openssl 1.1.0 API backward incompatibility.
I will try upstream version 11.1. Thank you anyhow.
